File: static/commands/duel.js
// static/commands/duel.js
(function () {
    const ACTIONS = {
        light: {
            key: "light",
            label: "Attaque rapide",
            desc: "Dégâts moyens, peu coûteux.",
            cost: 1,
            icon: "⚔️",
        },
        heavy: {
            key: "heavy",
            label: "Attaque puissante",
            desc: "Gros dégâts, peut rater.",
            cost: 2,
            icon: "💥",
        },
        heal: {
            key: "heal",
            label: "Soin",
            desc: "Récupère des PV.",
            cost: 2,
            icon: "✨",
        },
        shield: {
            key: "shield",
            label: "Bouclier",
            desc: "Réduit les dégâts reçus ce tour.",
            cost: 1,
            icon: "🛡️",
        },
        charge: {
            key: "charge",
            label: "Recharge",
            desc: "Ne fait rien ce tour mais +2 énergie.",
            cost: 0,
            icon: "⚡",
        },
        counter: {
            key: "counter",
            label: "Contre-attaque",
            desc: "Si l'autre attaque, ses dégâts se retournent contre lui. Sinon tu te blesses.",
            cost: 2,
            icon: "🔁",
        },
    };

    const state = {
        gameId: null,
        me: null,
        opponent: null,
        players: [],
        hp: {},
        energy: {},
        maxHp: 100,
        maxEnergy: 5,
        turn: 1,
        finished: false,
        waitingForResolution: false,
    };

    function resetState() {
        state.gameId = null;
        state.me = null;
        state.opponent = null;
        state.players = [];
        state.hp = {};
        state.energy = {};
        state.maxHp = 100;
        state.maxEnergy = 5;
        state.turn = 1;
        state.finished = false;
        state.waitingForResolution = false;
    }

    function setStateFromStart(payload, me) {
        state.gameId = payload.game_id;
        state.players = payload.players || [];
        state.hp = payload.hp || {};
        state.energy = payload.energy || {};
        state.maxHp = payload.max_hp || state.maxHp;
        state.maxEnergy = payload.max_energy || state.maxEnergy;
        state.turn = payload.turn || 1;
        state.me = me;
        state.opponent = state.players.find((p) => p !== me) || null;
        state.finished = false;
        state.waitingForResolution = false;
    }

    function renderInvite(ctx, payload, me) {
        resetState();
        state.gameId = payload.game_id;
        state.me = me;
        state.players = [payload.from, payload.to];
        state.hp = payload.hp || {};
        state.energy = payload.energy || {};
        state.maxHp = payload.max_hp || state.maxHp;
        state.maxEnergy = payload.max_energy || state.maxEnergy;
        state.opponent = me === payload.to ? payload.from : payload.to;
        state.turn = 1;
        state.finished = false;
        state.waitingForResolution = false;

        const isTarget = me === payload.to;
        const title = "Duel RPG";
        const subtitle = isTarget
            ? `${payload.from} te défie en duel !`
            : `Invitation envoyée à ${payload.to}…`;

        ctx.openModal(title, subtitle);

        const hpInfo = `
            <p class="text-xs text-slate-500">
                PV de départ : <span class="font-semibold">${state.maxHp}</span><br>
                Énergie max : <span class="font-semibold">${state.maxEnergy}</span>
            </p>
        `;

        const actionsInfo = `
            <ul class="text-xs text-slate-500 list-disc list-inside space-y-1">
                <li>⚔️ Attaque rapide : dégâts moyens, 1 énergie</li>
                <li>💥 Attaque puissante : gros dégâts mais peut rater, 2 énergies</li>
                <li>✨ Soin : récupère des PV, 2 énergies</li>
                <li>🛡️ Bouclier : réduit les dégâts ce tour, 1 énergie</li>
                <li>⚡ Recharge : +2 énergie, 0 énergie</li>
                <li>🔁 Contre-attaque : si l'autre t'attaque, ses dégâts se retournent contre lui. Sinon tu prends des dégâts.</li>
            </ul>
        `;

        if (isTarget) {
            ctx.setModalContent(`
                <div class="space-y-3">
                    <p class="text-sm text-slate-700">
                        ${payload.from} te propose un mini duel RPG.
                    </p>
                    ${hpInfo}
                    ${actionsInfo}
                </div>
            `);
            ctx.setModalActions(`
                <div class="flex justify-end gap-2">
                    <button id="duel-decline" class="px-3 py-1.5 text-xs rounded-lg border border-slate-200 text-slate-700 hover:bg-slate-50">
                        Refuser
                    </button>
                    <button id="duel-accept" class="px-3 py-1.5 text-xs rounded-lg bg-emerald-600 text-white hover:bg-emerald-700">
                        Accepter
                    </button>
                </div>
            `);

            const btnAccept = document.getElementById("duel-accept");
            const btnDecline = document.getElementById("duel-decline");

            if (btnAccept) {
                btnAccept.addEventListener("click", () => {
                    ctx.sendEvent("duel", "accept", {
                        game_id: state.gameId,
                    });
                });
            }

            if (btnDecline) {
                btnDecline.addEventListener("click", () => {
                    ctx.sendEvent("duel", "decline", {
                        game_id: state.gameId,
                    });
                    ctx.closeModal();
                    resetState();
                });
            }
        } else {
            ctx.setModalContent(`
                <div class="space-y-3">
                    <p class="text-sm text-slate-700">
                        En attente de la réponse de <span class="font-semibold">${payload.to}</span>…
                    </p>
                    ${hpInfo}
                    ${actionsInfo}
                </div>
            `);
            ctx.setModalActions(`
                <div class="flex justify-end">
                    <button id="duel-cancel" class="px-3 py-1.5 text-xs rounded-lg border border-slate-200 text-slate-700 hover:bg-slate-50">
                        Annuler
                    </button>
                </div>
            `);

            const btnCancel = document.getElementById("duel-cancel");
            if (btnCancel) {
                btnCancel.addEventListener("click", () => {
                    ctx.sendEvent("duel", "cancel", {
                        game_id: state.gameId,
                    });
                    ctx.closeModal();
                    resetState();
                });
            }
        }
    }

    function renderHpBar(current, max) {
        const pct = Math.max(0, Math.min(100, (current / max) * 100));
        return `
            <div class="w-full h-2 rounded-full bg-slate-200/80 overflow-hidden">
                <div class="h-2 rounded-full bg-emerald-500" style="width:${pct}%;"></div>
            </div>
        `;
    }

    function renderActionScreen(ctx) {
        const me = state.me;
        const opponent = state.opponent || "Adversaire";
        const hpMe = state.hp[me] ?? state.maxHp;
        const hpOpp = state.hp[opponent] ?? state.maxHp;
        const enMe = state.energy[me] ?? 0;
        const enOpp = state.energy[opponent] ?? 0;

        state.finished = false;
        state.waitingForResolution = false;

        ctx.openModal(
            "Duel RPG",
            `Tour ${state.turn} · ${me} vs ${opponent}`
        );

        let actionsHtml = "";
        Object.values(ACTIONS).forEach((a) => {
            actionsHtml += `
                <button
                    data-action="${a.key}"
                    class="duel-action flex-1 min-w-[120px] px-3 py-2 rounded-xl border border-slate-200 bg-white hover:bg-emerald-50 text-xs text-left space-y-1 transition"
                >
                    <div class="flex items-center justify-between gap-2">
                        <span class="text-lg">${a.icon}</span>
                        <span class="font-semibold text-slate-800 flex-1 truncate">
                            ${a.label}
                        </span>
                        <span class="text-[11px] text-slate-500">
                            ${a.cost} ⚡
                        </span>
                    </div>
                    <p class="text-[11px] text-slate-500">
                        ${a.desc}
                    </p>
                </button>
            `;
        });

        ctx.setModalContent(`
            <div class="space-y-4">
                <div class="grid grid-cols-2 gap-4 text-xs text-slate-600">
                    <div class="space-y-1">
                        <div class="flex items-center justify-between">
                            <span class="font-semibold text-slate-800">${me}</span>
                            <span>${hpMe}/${state.maxHp} PV</span>
                        </div>
                        ${renderHpBar(hpMe, state.maxHp)}
                        <div class="mt-1 text-[11px] text-slate-500">
                            Énergie : <span class="font-semibold">${enMe}</span> / ${state.maxEnergy}
                        </div>
                    </div>
                    <div class="space-y-1 text-right">
                        <div class="flex items-center justify-between">
                            <span>${hpOpp}/${state.maxHp} PV</span>
                            <span class="font-semibold text-slate-800">${opponent}</span>
                        </div>
                        ${renderHpBar(hpOpp, state.maxHp)}
                        <div class="mt-1 text-[11px] text-slate-500">
                            Énergie : <span class="font-semibold">${enOpp}</span> / ${state.maxEnergy}
                        </div>
                    </div>
                </div>

                <div class="space-y-2">
                    <p class="text-[11px] text-slate-500 text-center">
                        Choisis ton action pour ce tour :
                    </p>
                    <div class="flex flex-wrap gap-2 justify-center">
                        ${actionsHtml}
                    </div>
                </div>

                <p id="duel-status" class="text-[11px] text-center text-slate-500">
                    À toi de jouer.
                </p>
            </div>
        `);

        ctx.setModalActions(`
            <div class="flex justify-end">
                <button id="duel-cancel" class="px-3 py-1.5 text-xs rounded-lg border border-slate-200 text-slate-700 hover:bg-slate-50">
                    Abandonner
                </button>
            </div>
        `);

        const statusEl = document.getElementById("duel-status");
        const btnCancel = document.getElementById("duel-cancel");
        const buttons = Array.from(
            document.querySelectorAll(".duel-action")
        );

        if (btnCancel) {
            btnCancel.addEventListener("click", () => {
                if (state.finished) {
                    ctx.closeModal();
                    resetState();
                    return;
                }
                ctx.sendEvent("duel", "cancel", {
                    game_id: state.gameId,
                });
                ctx.closeModal();
                resetState();
            });
        }

        buttons.forEach((btn) => {
            const actionKey = btn.getAttribute("data-action");
            const a = ACTIONS[actionKey];
            if (!a) return;

            btn.addEventListener("click", () => {
                if (state.finished || state.waitingForResolution) return;

                const currentEnergy = state.energy[me] ?? 0;
                if (currentEnergy < a.cost) {
                    if (typeof ctx.toast === "function") {
                        ctx.toast(
                            `Tu n'as pas assez d'énergie pour ${a.label} (coût ${a.cost}).`,
                            "error"
                        );
                    }
                    if (statusEl) {
                        statusEl.textContent = "Pas assez d'énergie pour cette action.";
                    }
                    return;
                }

                state.waitingForResolution = true;

                buttons.forEach((b) => {
                    b.classList.add("opacity-60");
                    b.classList.remove("hover:bg-emerald-50");
                    b.disabled = true;
                });
                btn.classList.add("ring-2", "ring-emerald-400");

                if (statusEl) {
                    statusEl.textContent =
                        "Action envoyée. En attente de l'autre joueur…";
                }

                ctx.sendEvent("duel", "action", {
                    game_id: state.gameId,
                    action: actionKey,
                });
            });
        });
    }

    function renderRoundResult(ctx, payload) {
        const turn = payload.turn || state.turn;
        const results = payload.results || [];
        const hp = payload.hp || state.hp;
        const energy = payload.energy || state.energy;
        const winner = payload.winner || null;
        const draw = !!payload.draw;
        const finished = !!payload.finished;

        state.hp = hp;
        state.energy = energy;
        state.turn = turn + (finished ? 0 : 1);
        state.finished = finished;
        state.waitingForResolution = false;

        const me = state.me;
        const opponent = state.opponent || "Adversaire";

        const rMe = results.find((r) => r.player === me) || {};
        const rOpp = results.find((r) => r.player === opponent) || {};

        function actionEmoji(res) {
            const a = ACTIONS[res.action];
            return a ? a.icon : "❔";
        }

        function actionLabel(res) {
            if (!res.action) return "Aucune action";
            const a = ACTIONS[res.action];
            return a ? a.label : res.action;
        }

        function resultText(res) {
            if (res.result === "hit") {
                return `touche et inflige ${res.damage_dealt || 0} dégâts.`;
            }
            if (res.result === "miss") {
                return "rate son attaque.";
            }
            if (res.result === "heal") {
                return `se soigne de ${res.healed || 0} PV.`;
            }
            if (res.result === "shield") {
                return "se protège derrière un bouclier.";
            }
            if (res.result === "charge") {
                return "se concentre et recharge son énergie.";
            }
            if (res.result === "counter_success") {
                return "réussit sa contre-attaque : les dégâts se retournent contre l'adversaire.";
            }
            if (res.result === "counter_fail") {
                return "tente une contre-attaque dans le vide et se blesse.";
            }
            if (res.result === "no_energy") {
                return "n'avait pas assez d'énergie, l'action échoue.";
            }
            return "ne fait rien de spécial.";
        }

        let globalLine = "";
        if (draw) {
            globalLine = `Égalité ! Les deux combattants tombent en même temps…`;
        } else if (winner) {
            globalLine = `${winner} met fin au duel à ce tour !`;
        } else {
            globalLine = `Le combat continue…`;
        }

        ctx.openModal(
            "Duel RPG",
            `Résultat du tour ${turn}`
        );

        ctx.setModalContent(`
            <div class="space-y-4">
                <div class="grid grid-cols-2 gap-4 text-xs text-slate-600">
                    <div class="space-y-1">
                        <div class="flex items-center justify-between">
                            <span class="font-semibold text-slate-800">${me}</span>
                            <span>${hp[me] ?? 0}/${state.maxHp} PV</span>
                        </div>
                        ${renderHpBar(hp[me] ?? 0, state.maxHp)}
                        <div class="mt-1 text-[11px] text-slate-500">
                            Énergie : <span class="font-semibold">${energy[me] ?? 0}</span> / ${state.maxEnergy}
                        </div>
                    </div>
                    <div class="space-y-1 text-right">
                        <div class="flex items-center justify-between">
                            <span>${hp[opponent] ?? 0}/${state.maxHp} PV</span>
                            <span class="font-semibold text-slate-800">${opponent}</span>
                        </div>
                        ${renderHpBar(hp[opponent] ?? 0, state.maxHp)}
                        <div class="mt-1 text-[11px] text-slate-500">
                            Énergie : <span class="font-semibold">${energy[opponent] ?? 0}</span> / ${state.maxEnergy}
                        </div>
                    </div>
                </div>

                <div class="flex items-center justify-center gap-8 text-3xl md:text-4xl">
                    <div class="flex flex-col items-center gap-1">
                        <span>${actionEmoji(rMe)}</span>
                        <span class="text-[11px] text-slate-500">${me}</span>
                    </div>
                    <div class="text-xl text-slate-400">vs</div>
                    <div class="flex flex-col items-center gap-1">
                        <span>${actionEmoji(rOpp)}</span>
                        <span class="text-[11px] text-slate-500">${opponent}</span>
                    </div>
                </div>

                <div class="space-y-1 text-xs text-slate-700">
                    <p><span class="font-semibold">${me}</span> ${actionLabel(rMe)} et ${resultText(rMe)}</p>
                    <p><span class="font-semibold">${opponent}</span> ${actionLabel(rOpp)} et ${resultText(rOpp)}</p>
                </div>

                <div class="text-center text-xs text-slate-600">
                    ${globalLine}
                </div>
            </div>
        `);

        if (finished) {
            ctx.setModalActions(`
                <div class="flex justify-end">
                    <button id="duel-close" class="px-3 py-1.5 text-xs rounded-lg bg-emerald-600 text-white hover:bg-emerald-700">
                        Fermer
                    </button>
                </div>
            `);

            const btnClose = document.getElementById("duel-close");
            if (btnClose) {
                btnClose.addEventListener("click", () => {
                    ctx.closeModal();
                    resetState();
                });
            }
        } else {
            ctx.setModalActions(`
                <div class="flex justify-end">
                    <button id="duel-next" class="px-3 py-1.5 text-xs rounded-lg bg-emerald-600 text-white hover:bg-emerald-700">
                        Tour suivant
                    </button>
                </div>
            `);

            const btnNext = document.getElementById("duel-next");
            if (btnNext) {
                btnNext.addEventListener("click", () => {
                    renderActionScreen(ctx);
                });
            }
        }
    }

    function renderMatchEnd(ctx, payload) {
        const winner = payload.winner || null;
        const draw = !!payload.draw;
        const hp = payload.hp || state.hp;
        const players = payload.players || state.players;

        state.hp = hp;
        state.finished = true;

        const p1 = players[0] || "";
        const p2 = players[1] || "";

        let title;
        if (draw) {
            title = "Égalité !";
        } else if (winner) {
            title = `${winner} remporte le duel !`;
        } else {
            title = "Duel terminé";
        }

        ctx.openModal("Duel RPG", title);

        ctx.setModalContent(`
            <div class="space-y-3 text-xs text-slate-600">
                <div class="flex items-center justify-between">
                    <div>
                        <p class="font-semibold text-slate-800">${p1}</p>
                        <p>PV restants : ${hp[p1] ?? 0}/${state.maxHp}</p>
                    </div>
                    <div class="text-right">
                        <p class="font-semibold text-slate-800">${p2}</p>
                        <p>PV restants : ${hp[p2] ?? 0}/${state.maxHp}</p>
                    </div>
                </div>
                <p class="text-center text-[11px] text-slate-500">
                    Merci d'avoir joué !
                </p>
            </div>
        `);

        ctx.setModalActions(`
            <div class="flex justify-end">
                <button id="duel-close" class="px-3 py-1.5 text-xs rounded-lg bg-emerald-600 text-white hover:bg-emerald-700">
                    Fermer
                </button>
            </div>
        `);

        const btnClose = document.getElementById("duel-close");
        if (btnClose) {
            btnClose.addEventListener("click", () => {
                ctx.closeModal();
                resetState();
            });
        }
    }

    function renderCanceled(ctx, payload) {
        const reason = payload.reason || "Le duel a été annulé.";
        if (typeof ctx.toast === "function") {
            ctx.toast(reason, "info");
        }
        ctx.closeModal();
        resetState();
    }

    // Enregistrement du client de commande "duel"
    window.registerCommandClient("duel", {
        onServerMessage(data, ctx) {
            const payload = data.payload || {};
            const me = ctx.getCurrentUser();

            if (!payload.type) return;

            if (
                payload.game_id &&
                state.gameId &&
                payload.game_id !== state.gameId
            ) {
                return;
            }

            if (payload.type === "invite") {
                renderInvite(ctx, payload, me);
                return;
            }

            if (payload.type === "start") {
                setStateFromStart(payload, me);
                renderActionScreen(ctx);
                return;
            }

            if (payload.type === "round_result") {
                renderRoundResult(ctx, payload);
                return;
            }

            if (payload.type === "match_end") {
                renderMatchEnd(ctx, payload);
                return;
            }

            if (payload.type === "canceled") {
                renderCanceled(ctx, payload);
                return;
            }
        },
    });
})();

File: commands/duel.py
import uuid
import random
from datetime import datetime, timedelta


GAMES = {}

MAX_HP = 100
MAX_ENERGY = 5

ACTIONS = {
    "light": {
        "label": "Attaque rapide",
        "cost": 1,
        "min_dmg": 10,
        "max_dmg": 18,
        "hit_chance": 0.95,
    },
    "heavy": {
        "label": "Attaque puissante",
        "cost": 2,
        "min_dmg": 22,
        "max_dmg": 35,
        "hit_chance": 0.7,
    },
    "heal": {
        "label": "Soin",
        "cost": 2,
        "min_heal": 12,
        "max_heal": 24,
    },
    "shield": {
        "label": "Bouclier",
        "cost": 1,
    },
    "charge": {
        "label": "Recharge",
        "cost": 0,
        "energy_gain": 2,
    },
    "counter": {
        "label": "Contre-attaque",
        "cost": 2,
    },
}


def _now():
    return datetime.utcnow()


def _cleanup_expired():
    expiration = _now() - timedelta(hours=1)
    to_delete = []
    for gid, g in GAMES.items():
        if g.get("created_at") and g["created_at"] < expiration:
            to_delete.append(gid)
        elif g.get("status") == "finished":
            to_delete.append(gid)
    for gid in to_delete:
        GAMES.pop(gid, None)


def _find_user(u_col, username: str):
    return u_col.find_one({"nom": username})


def _create_game(ctx, opponent_name: str):
    host = ctx.username
    opponent = opponent_name
    players = [host, opponent]
    game_id = str(uuid.uuid4())

    hp = {host: MAX_HP, opponent: MAX_HP}

    energy = {host: 3, opponent: 3}

    GAMES[game_id] = {
        "id": game_id,
        "channel_key": ctx.channel_key,
        "channel_type": ctx.channel_type,
        "players": players,
        "hp": hp,
        "energy": energy,
        "choices": {host: None, opponent: None},
        "status": "pending",
        "turn": 1,
        "created_at": _now(),
    }

    payload = {
        "type": "invite",
        "game_id": game_id,
        "from": host,
        "to": opponent,
        "hp": hp,
        "energy": energy,
        "max_hp": MAX_HP,
        "max_energy": MAX_ENERGY,
    }
    ctx.send_command_to(players, "duel", payload)


def handle(ctx, args):
    _cleanup_expired()

    if not args:
        ctx.send_to_self(
            {
                "type": "error",
                "message": "Utilisation : /duel pseudo_adversaire",
            }
        )
        return

    opponent = args[0].strip()
    if opponent.startswith("@"):
        opponent = opponent[1:]

    if not opponent:
        ctx.send_to_self(
            {
                "type": "error",
                "message": "Tu dois spécifier un pseudo.",
            }
        )
        return

    if opponent == ctx.username:
        ctx.send_to_self(
            {
                "type": "error",
                "message": "Tu ne peux pas te défier toi-même.",
            }
        )
        return

    target_user = _find_user(ctx.u_col, opponent)
    if not target_user:
        ctx.send_to_self(
            {
                "type": "error",
                "message": f"Utilisateur introuvable : {opponent }.",
            }
        )
        return

    for g in GAMES.values():
        if g["channel_key"] == ctx.channel_key and g["status"] in (
            "pending",
            "playing",
        ):
            if ctx.username in g["players"] or opponent in g["players"]:
                ctx.send_to_self(
                    {
                        "type": "error",
                        "message": "Un duel est déjà en cours ou en attente entre ces joueurs.",
                    }
                )
                return

    _create_game(ctx, opponent)
    ctx.send_to_self(
        {
            "type": "info",
            "message": f"Invitation à un duel envoyée à {opponent }.",
        }
    )


def _get_game_or_error(ctx, payload):
    game_id = (payload or {}).get("game_id")
    if not game_id or game_id not in GAMES:
        ctx.send_to_self(
            {
                "type": "error",
                "message": "Duel introuvable ou expiré.",
            }
        )
        return None
    game = GAMES[game_id]
    if ctx.username not in game["players"]:
        ctx.send_to_self(
            {
                "type": "error",
                "message": "Tu ne fais pas partie de ce duel.",
            }
        )
        return None
    return game


def _send_start(ctx, game):
    payload = {
        "type": "start",
        "game_id": game["id"],
        "players": game["players"],
        "hp": game["hp"],
        "energy": game["energy"],
        "max_hp": MAX_HP,
        "max_energy": MAX_ENERGY,
        "turn": game["turn"],
    }
    ctx.send_command_to(game["players"], "duel", payload)


def _resolve_turn(ctx, game):
    players = game["players"]
    p1, p2 = players[0], players[1]
    hp = game["hp"]
    energy = game["energy"]
    choices = game["choices"]
    turn = game["turn"]

    energy_regen = {p: energy[p] for p in players}

    actions_data = {}
    pending_damage = {p1: [], p2: []}
    damage_dealt = {p1: 0, p2: 0}
    damage_taken = {p1: 0, p2: 0}
    heal_done = {p1: 0, p2: 0}
    counter_flags = {p1: False, p2: False}
    self_damage_from_counter_fail = {p1: 0, p2: 0}

    for p in players:
        other = p2 if p == p1 else p1
        action_key = choices.get(p)
        spec = ACTIONS.get(action_key) if action_key else None

        if not spec:
            actions_data[p] = {
                "action": None,
                "label": "Rien",
                "cost": 0,
                "enough": True,
                "result": "none",
                "energy_after": energy_regen[p],
                "shield": False,
            }
            continue

        cost = spec["cost"]
        energy_after = energy_regen[p]
        enough = energy_after >= cost

        result_type = None
        shield_active = False

        if not enough:

            result_type = "no_energy"
        else:

            energy_after -= cost

            if action_key in ("light", "heavy"):
                hit_chance = spec["hit_chance"]
                if random.random() <= hit_chance:
                    dmg = random.randint(spec["min_dmg"], spec["max_dmg"])
                    pending_damage[other].append((dmg, p))
                    result_type = "hit"
                else:
                    result_type = "miss"

            elif action_key == "heal":
                heal = random.randint(spec["min_heal"], spec["max_heal"])
                heal_done[p] += heal
                result_type = "heal"

            elif action_key == "shield":
                shield_active = True
                result_type = "shield"

            elif action_key == "charge":
                gain = spec.get("energy_gain", 2)
                energy_after = min(MAX_ENERGY, energy_after + gain)
                result_type = "charge"

            elif action_key == "counter":
                counter_flags[p] = True
                result_type = "counter_wait"

        actions_data[p] = {
            "action": action_key,
            "label": spec["label"],
            "cost": cost,
            "enough": enough,
            "result": result_type,
            "energy_after": energy_after,
            "shield": shield_active,
        }

    for p in players:
        if not counter_flags[p]:
            continue
        if not actions_data[p]["enough"]:
            continue

        other = p2 if p == p1 else p1

        hits = pending_damage[p]
        if hits:

            total_reflected = sum(d for d, _ in hits)
            pending_damage[p] = []
            pending_damage[other].append((total_reflected, p))

            actions_data[p]["result"] = "counter_success"
            actions_data[p]["reflected"] = total_reflected
        else:

            fail_dmg = random.randint(8, 16)
            self_damage_from_counter_fail[p] += fail_dmg
            actions_data[p]["result"] = "counter_fail"
            actions_data[p]["self_damage"] = fail_dmg

    delta_hp_total = {p1: 0, p2: 0}

    for p in players:
        delta_hp_total[p] += heal_done[p]

    for p in players:
        if self_damage_from_counter_fail[p] > 0:
            dmg = self_damage_from_counter_fail[p]
            delta_hp_total[p] -= dmg
            damage_taken[p] += dmg

    for target in players:
        shielded = actions_data[target]["shield"]
        for dmg_raw, source in pending_damage[target]:
            if shielded:
                dmg_final = max(1, int(dmg_raw * 0.5))
            else:
                dmg_final = dmg_raw

            delta_hp_total[target] -= dmg_final
            damage_taken[target] += dmg_final
            damage_dealt[source] += dmg_final

    new_hp = {}
    for p in players:
        val = hp[p] + delta_hp_total[p]
        if val > MAX_HP:
            val = MAX_HP
        if val < 0:
            val = 0
        new_hp[p] = val

    new_energy = {p: actions_data[p]["energy_after"] for p in players}

    winner_player = None
    draw = False

    if new_hp[p1] <= 0 and new_hp[p2] <= 0:
        draw = True
    elif new_hp[p1] <= 0:
        winner_player = p2
    elif new_hp[p2] <= 0:
        winner_player = p1

    finished = bool(winner_player or draw)

    results = []
    for p in players:
        results.append(
            {
                "player": p,
                "hp_before": hp[p],
                "hp_after": new_hp[p],
                "energy_before": energy[p],
                "energy_after": new_energy[p],
                "action": actions_data[p]["action"],
                "label": actions_data[p]["label"],
                "result": actions_data[p]["result"],
                "damage_dealt": damage_dealt[p],
                "damage_taken": damage_taken[p],
                "healed": heal_done[p],
            }
        )

    ctx.send_command_to(
        players,
        "duel",
        {
            "type": "round_result",
            "game_id": game["id"],
            "turn": turn,
            "hp": new_hp,
            "energy": new_energy,
            "max_hp": MAX_HP,
            "max_energy": MAX_ENERGY,
            "results": results,
            "winner": winner_player,
            "draw": draw,
            "finished": finished,
        },
    )

    if finished:
        ctx.send_command_to(
            players,
            "duel",
            {
                "type": "match_end",
                "game_id": game["id"],
                "winner": winner_player,
                "hp": new_hp,
                "energy": new_energy,
                "players": players,
                "draw": draw,
            },
        )
        game["status"] = "finished"
        GAMES.pop(game["id"], None)
    else:
        game["hp"] = new_hp
        game["energy"] = new_energy
        game["turn"] = turn + 1
        for p in players:
            game["choices"][p] = None


def handle_event(ctx, event_name: str, payload: dict):
    _cleanup_expired()
    event_name = (event_name or "").strip().lower()

    if event_name not in ("accept", "decline", "cancel", "action"):
        ctx.send_to_self(
            {
                "type": "error",
                "message": "Action de duel invalide.",
            }
        )
        return

    game = _get_game_or_error(ctx, payload)
    if not game:
        return

    players = game["players"]
    p1, p2 = players[0], players[1]

    if event_name == "accept":
        invited = p2
        if ctx.username != invited:
            ctx.send_to_self(
                {
                    "type": "error",
                    "message": "Seul l'adversaire peut accepter le duel.",
                }
            )
            return
        if game["status"] != "pending":
            ctx.send_to_self(
                {
                    "type": "error",
                    "message": "Ce duel a déjà été accepté ou annulé.",
                }
            )
            return
        game["status"] = "playing"
        _send_start(ctx, game)
        return

    if event_name == "decline":
        invited = p2
        if ctx.username != invited:
            ctx.send_to_self(
                {
                    "type": "error",
                    "message": "Seul l'adversaire peut refuser le duel.",
                }
            )
            return
        if game["status"] != "pending":
            ctx.send_to_self(
                {
                    "type": "error",
                    "message": "Ce duel a déjà été accepté ou annulé.",
                }
            )
            return
        game["status"] = "finished"
        ctx.send_command_to(
            players,
            "duel",
            {
                "type": "canceled",
                "game_id": game["id"],
                "reason": f"{ctx .username } a refusé le duel.",
            },
        )
        GAMES.pop(game["id"], None)
        return

    if event_name == "cancel":
        if game["status"] not in ("pending", "playing"):
            ctx.send_to_self(
                {
                    "type": "error",
                    "message": "Le duel est déjà terminé.",
                }
            )
            return

        game["status"] = "finished"
        ctx.send_command_to(
            players,
            "duel",
            {
                "type": "canceled",
                "game_id": game["id"],
                "reason": f"{ctx .username } a annulé le duel.",
            },
        )
        GAMES.pop(game["id"], None)
        return

    if event_name == "action":
        if game["status"] != "playing":
            ctx.send_to_self(
                {
                    "type": "error",
                    "message": "Le duel n'est pas en cours.",
                }
            )
            return

        action_key = (payload or {}).get("action")
        if not isinstance(action_key, str) or action_key not in ACTIONS:
            ctx.send_to_self(
                {
                    "type": "error",
                    "message": "Action inconnue.",
                }
            )
            return

        if game["choices"][ctx.username] is not None:
            ctx.send_to_self(
                {
                    "type": "error",
                    "message": "Tu as déjà choisi ton action pour ce tour.",
                }
            )
            return

        game["choices"][ctx.username] = action_key

        if all(game["choices"].get(p) is not None for p in players):
            _resolve_turn(ctx, game)
        else:
            ctx.send_to_self(
                {
                    "type": "info",
                    "message": "Action enregistrée, en attente de l'autre joueur.",
                }
            )
